tools/autograd/derivatives.yaml

```diff
@@ -121,6 +121,11 @@
   tensor1: grad * tensor2 * value
   tensor2: grad * tensor1 * value
 
+- name: linear(Tensor self, Tensor weight, Tensor bias)
```

Comment:
I don't see how this can work -- isn't this assuming that things are 2-d?

Reply:
Better solutions would seem to be:
1) If mkldnn supports them, implement addmm, matmul etc.
2) If not, do the dispatch in the op.

Reply:
For cpu and cuda, the aten linear is exactly the same as the python (i.e. dispatch to addmm is it's 2d and has bias, otherwise use matmul). I think I just need to add assertion in the mkldnn_linear to say right now we only support 2d linear.

Reply:
I would like to avoid dispatching in the op as much as possible, because that will end up with a lot of places doing dispatch (potentially in different ways).
For this one, mkldnn supports linear, but not addmm, and to support nn.Linear, there is no need to support the whole widened semantics of addmm. The change in this PR doesn't change anything in the cpu and cuda sides, it just make other libraries to support nn.Linear easier.

Reply:
We have code in ideep `inner_product_forward` to handle the case when input tensor is >2D. No extra assertion is needed.
https://github.com/intel/ideep/blob/960ee3a20d975655d3fa8e40e445898a1e3b4078/include/ideep/computations.hpp#L5450

Reply:
> I would like to avoid dispatching in the op as much as possible, because that will end up with a lot of places doing dispatch (potentially in different ways).

As I tried to express in https://github.com/pytorch/pytorch/pull/19204, this is not a good approach.  For one, it makes calculating the derivatives much more difficult (as seen here).  For two, it's difficult to keep the semantics of the functions in sync because the code lives in completely separate places (as demonstrated in the conv2d patch).  There's just no good way around actually understanding what is in the existing function -- putting everything in a separate dispatch and pretending these are standalone functions is going to be a bad user experience when these functions ultimately diverge.

Reply:
On this topic - why can't we just move python's functionl.nn.Linear implementation to C++ and do dispatch there? It always bothered me that for such a critical op we effectively produce different ops depending on bias vs no bias.

Reply:
It depends on what you mean by "op".  The derivative issue is there whether or not we are in C++.

Reply:
we could write derivative for linear() explicitly too. Given it's such a popular op, I'd rather make it atomic